tools/jit/templates/aten_dispatch.cpp

```diff
@@ -1,4 +1,3 @@
-#include "Python.h"
```

Comment:
Have you tried building with Python 2? Changes like this one might cause some "macro redefined" warnings to appear

Reply:
Yep, I've seen the warnings popping up.
The workaround I'd like to propose is the following: we create a `python_header.h` (suggestions for naming are welcome) in `torch/csrc/utils` containing
```
#undef _POSIX_C_SOURCE
#undef _POSIX_THREADS
#undef _XOPEN_SOURCE

#include <Python.h>
```
and include that file instead of including `Python.h` directly throughout the codebase. This should allow us to remove all `#include <Python.h>` where it's not needed.

This has been done in other projects, e.g. https://github.com/Slicer/VTK/blob/58e382c061ec870de18ef1516db3009c30ae6577/Wrapping/Python/vtkPython.h

Reply:
I'm not convinced this is a good idea.

In an ideal world, shouldn't we end up with an include structure where non-Python-using headers never include `Python.h` (transitively), and Python-using headers include `Python.h` at their header. In this case, shouldn't this workaround not be necessary?

Reply:
So, as I understand it, the issue is that non-Python-using headers currently include `Python.h` only to avoid the warnings, just because `Python.h` (version 2.x) doesn't `undef` 3 symbols before (re-)defining them.

The solution I've seen floating around is to manually undef those symbols before including `Python.h` in Python-using headers.

In this PR, I've removed all unnecessary `Python.h` includes in non-Python-using headers.
The last part of the fix would be to create a new header where we just `undef` those symbols and include `Python.h`. The idea would be to include that header instead of including `Python.h` directly in Python-using-headers.

The (maybe) more explicit alternative would be to create a macro that undefs those symbols to call before including `Python.h` in Python-using-headers. I'm not sure that's any better.

I hope I'm getting it in general :-)

Reply:
My point is that you only need to do this if the header transitively includes a header that includes `Python.h`. Which, non-Python using headers should not! (If they do, that indicates we have designed our headers wrong.)

Reply:
Are we sure that these symbols are not defined by other headers (e.g. GNU libc)? My impression was that they are, and that `Python.h` is just not a good citizen and assumes nobody does, so it doesn't `undef` them. But I could be wrong, I can research it more.

Reply:
Python is *absolutely* in the wrong for defining these macros, but undef'ing `_XOPEN_SOURCE` is even more wrong. The net effect is that you'll end up with a value for `_XOPEN_SOURCE` which doesn't actually reflect how the headers that you have actually included were interpreted.

Reply:
So basically our goal is to make sure we include one file that includes `Python.h` before everything else in the Python build? (the problem is just not there for the cpp build)

Reply:
Yes. I think an easy to apply heuristic is to put `#include "Python.h"` at the beginning of every Python-using header file.

I am sure you will run into problems making this happen, but this is a good place to start.

Reply:
Yep, which is what happens now in this PR.
The thing is: if any non Python-using header is included first, then those symbols get defined in some c++ header and we get the warning as soon as we hit the first Python.h include.
So we must ensure that Python.h is the first thing to be included in a python-enabled build.

Reply:
Uh, looks like it was easier than expected!